Corona-Warn-App/src/main/java/de/rki/coronawarnapp/util/serialization/adapter/ByteArrayAdapter.kt

```diff
@@ -0,0 +1,20 @@
+package de.rki.coronawarnapp.util.serialization.adapter
+
+import com.google.gson.TypeAdapter
+import com.google.gson.stream.JsonReader
+import com.google.gson.stream.JsonWriter
+import okio.ByteString.Companion.decodeBase64
+import okio.ByteString.Companion.toByteString
+import org.json.JSONObject.NULL
+
+class ByteArrayAdapter : TypeAdapter<ByteArray>() {
+    override fun write(out: JsonWriter, value: ByteArray?) {
+        if (value == null) out.nullValue()
+        else value.toByteString().base64().let { out.value(it) }
+    }
+
+    override fun read(reader: JsonReader): ByteArray? = when (reader.peek()) {
+        NULL -> reader.nextNull().let { null }
+        else -> reader.nextString().decodeBase64()!!.toByteArray()
```

Comment:
```suggestion
        else -> reader.nextString().decodeBase64()?.toByteArray() ?: byteArrayOf()
```

Reply:
alternativ:
```suggestion
        else -> reader.nextString().decodeBase64()?.toByteArray()
```

Reply:
This would change the returned data though, if we are not reading a NULL token, then we expect there to be valid data, we want to crash then, not return "null" as if there would be no data.

AFAICT (see [here](https://github.com/square/okio/blob/master/okio/src/commonMain/kotlin/okio/-Base64.kt#L31)), the only way for `decodeBase64` to return null is on invalid data. So if we then return null in the not `NULL` token case, then we treat not existing data and invalid data the same. Would that be desired behavior?

Reply:
ok, if a crash is ok with everyone - let's do it!

Reply:
I think the crash needs to happen, otherwise the classes above, using gson (with this adapter) to deserialize something, wouldn't know that there is something wrong with the data if there was no exception. What do you think?

Reply:
I believe this app should never crash in production. :( but my opinion is not the only one that counts.

Reply:
That's not what I meant.

The code we are talking about is part of a custom type adapter for data serialization via Gson.

The TypeAdapter does not have enough information to determine whether returning `null` or throwing an exception is the correct approach, so it should not make that decision, but throw an exception so that the "higher level classes" know about this issue and can make the decision as they have more information available.

```kotlin
    fun getLastTimeStampData() : Instant? {
        return try {
            gson.fromJson(someFile)
        } catch (e: Exception) {
            Timber.e(e, "Our data was corrupted, returning EPOCH")
            Instant.EPOCH
        }
    }
```

The `getLastTimeStampData` knows about the data's purpose and can make a decision whether we should return null in an error case, or a fallback value. If we catch the decoding issue in the `ByteArrayAdapter` and just return `null`, then `getLastTimeStampData` could no longer distinguish between "no value exists" and "value exists but data corrupted/invalid".

Reply:
Thinking about it, it would be good to throw a more specific Exception than a Kotlin NPE. Will add that.

Reply:
Let me add a unit test for before it's merged.